### PR TITLE
Fix repeating masks in masking training mode

### PR DESCRIPTION
**Pull request by kacpnowak**
## Description

Fixes issue with mask being repeated by each worker by setting TokenizerMasking rng based on worker's ID

## Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes https://github.com/ecmwf/WeatherGenerator/issues/267

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/weathergen/datasets/tokenizer_masking.py [hunk 1]
```diff
@@ -288,9 +292,11 @@ def id(arg):
         # select masked tokens for network input
         # TODO: implement sampling rate target
         target_tokens = [
-            torch.cat([c if p else torch.tensor([]) for c, p in zip(cc, pp, strict=True)])
-            if len(cc) > 0
-            else torch.tensor(cc)
+            (
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_masking.py:295]: There still seems to be a problem with your ruff settings

File: src/weathergen/datasets/tokenizer_masking.py [hunk 2]
```diff
@@ -130,7 +132,9 @@ def get_size_time_embedding(self) -> int:
         return self.size_time_embedding
 
     def reset(self) -> None:
-        self.rng = np.random.default_rng(int(time.time()))
+        worker_info = torch.utils.data.get_worker_info()
+        id = worker_info.id if worker_info is not None else 1
+        self.rng = np.random.default_rng(int(time.time()) + id)
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_masking.py]: What is the advantage compared to the version that I proposed in the issue?

**Review by clessig — COMMENTED** [review]
Many thanks for the quick fix. Just a minor question

File: src/weathergen/datasets/tokenizer_masking.py (hunk 2 above)
Comment (kacpnowak) [reply src/weathergen/datasets/tokenizer_masking.py]: In your case, seed was exactly the same for each worker. By using ID it's unique to each of them

File: src/weathergen/datasets/tokenizer_masking.py (hunk 1 above)
Comment (kacpnowak) [reply src/weathergen/datasets/tokenizer_masking.py:295]: I really don't know how it keeps happening. I'm always using `uv run --no-project --with "ruff==0.9.7" ruff format --check -n src/` before commiting

File: src/weathergen/datasets/tokenizer_masking.py (hunk 2 above)
Comment (clessig) [reply src/weathergen/datasets/tokenizer_masking.py]: But the bug was in `num_workers` here instead of `id`:

div_factor = worker_info.num_workers if worker_info is not None else 1

Why the change in the second line. I would prefer the version with div_factor

Comment (tjhunter) [reply src/weathergen/datasets/tokenizer_masking.py]: rather than trying some math formulas, how about a hash:

```py
seed = abs(hash((worker_info.id, time.time())))
self.rng = np.random.default_rng(seed=seed)
```

Also, I would rather we have an index incremented at each reset() rather than using `time()`. This would be deterministic.

Comment (tjhunter) [reply src/weathergen/datasets/tokenizer_masking.py]: (and this works for None by the way)

Comment (clessig) [reply src/weathergen/datasets/tokenizer_masking.py]: I prefer something random; you don't want to lose 5% because you always run the optimization/training along the same sub-optimal path. There are other sources of randomness but the masking is a critical part. We can ensure that problems are repeatable by providing the sample index in stream_info. 

No strong feeling on hash versus math :)

Comment (kacpnowak) [reply src/weathergen/datasets/tokenizer_masking.py]: > rather than trying some math formulas, how about a hash:
> 
> ```python
> seed = abs(hash((worker_info.id, time.time())))
> self.rng = np.random.default_rng(seed=seed)
> ```
> 
> Also, I would rather we have an index incremented at each reset() rather than using `time()`. This would be deterministic.

I think both solutions work just as good.

Just one thing needs fixing, accessing the `id` in case the `worker_info` in `None` will throw `AttributeError` so the entire object needs to be hashed

Comment (tjhunter) [reply src/weathergen/datasets/tokenizer_masking.py]: I would not hash the entire object and just the id (or None). the reason being that there is no guarantee that other subfields of the object would be themselves hashable or even stable between repeated calls.

Comment (kacpnowak) [reply src/weathergen/datasets/tokenizer_masking.py]: Okay, so if you and @clessig approve this solution I'll commit it

Comment (clessig) [reply src/weathergen/datasets/tokenizer_masking.py]: Can we use

```
div_factor = worker_info.id if worker_info is not None else 1
self.rng = np.random.default_rng(int(time.time() / div_factor))
```

I feel more comfortable with seeds that are not so clustered.
